test/e2e/framework/framework.go

```diff
@@ -0,0 +1,333 @@
+			if apierrors.IsNotFound(err) {
+				return true, nil
+			}
+			Logf("Error while waiting for namespace to be terminated: %v", err)
+			return false, nil
+		}
+		return false, nil
+	})
+
+	// Namespace deletion timed out.
+	if err != nil {
+		return fmt.Errorf("namespace %v was not deleted with limit: %v", namespace, err)
+	}
+
+	Logf("namespace %v deletion completed in %s", namespace, time.Now().Sub(startTime))
+	return nil
+}
+
+// BeforeEach gets a client and makes a namespace.
+func (f *Framework) BeforeEach() {
+	// The fact that we need this feels like a bug in ginkgo.
+	// https://github.com/onsi/ginkgo/issues/222
+	f.cleanupHandle = AddCleanupAction(f.AfterEach)
+
+	if f.ClientSet == nil {
+		By("Creating a kubernetes client")
+		config, err := clientcmd.BuildConfigFromFlags("", TestContext.KubeConfig)
+		Expect(err).NotTo(HaveOccurred())
+		f.ClientSet, err = clientset.NewForConfig(config)
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	if TestContext.Namespace == "" {
+		By("Building a namespace api object")
+		namespace, err := f.CreateNamespace(f.BaseName, map[string]string{
+			"e2e-framework": f.BaseName,
+		})
+		Expect(err).NotTo(HaveOccurred())
+		f.Namespace = namespace
+	} else {
+		By(fmt.Sprintf("Getting existing namespace %q", TestContext.Namespace))
+		namespace, err := f.ClientSet.CoreV1().Namespaces().Get(TestContext.Namespace, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		f.Namespace = namespace
+	}
+
+	if f.IsBackup {
+		f.BlockStorageClient = f.createStorageClient()
+	}
+
+	if !f.ProvisionerFSSInstalled {
+		err := f.installFSSProvisioner(f.Namespace.Name)
+		Expect(err).NotTo(HaveOccurred())
+		f.ProvisionerFSSInstalled = true
+	}
+
+	if !f.ProvisionerBlockInstalled {
+		err := f.installBlockProvisioner(f.Namespace.Name)
+		Expect(err).NotTo(HaveOccurred())
+		f.ProvisionerBlockInstalled = true
+	}
+}
+
+// AfterEach deletes the namespace(s).
+func (f *Framework) AfterEach() {
+	RemoveCleanupAction(f.cleanupHandle)
+
+	nsDeletionErrors := map[string]error{}
+
+	// Whether to delete namespace is determined by 3 factors: delete-namespace flag, delete-namespace-on-failure flag and the test result
+	// if delete-namespace set to false, namespace will always be preserved.
+	// if delete-namespace is true and delete-namespace-on-failure is false, namespace will be preserved if test failed.
+	if TestContext.DeleteNamespace && (TestContext.DeleteNamespaceOnFailure || !CurrentGinkgoTestDescription().Failed) {
+		for _, ns := range f.namespacesToDelete {
+			By(fmt.Sprintf("Destroying namespace %q for this suite.", ns.Name))
+			if err := f.DeleteNamespace(ns.Name, 5*time.Minute); err != nil {
+				nsDeletionErrors[ns.Name] = err
+			}
+		}
+	}
+
+	for _, backupID := range f.BackupIDs {
+		By(fmt.Sprintf("Deleting backups %q", backupID))
+		ctx := context.Background()
+		f.BlockStorageClient.DeleteVolumeBackup(ctx, coreOCI.DeleteVolumeBackupRequest{VolumeBackupId: &backupID})
```

Comment:
Should report on errors deleting volumes in the same way that we do for namespaces (the same will go for StorageClasses presuming we move their management to the framework).